flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataStreamProcTimeAggregateGlobalWindowFunction.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.aggregate
+
+import org.apache.flink.streaming.api.functions.windowing.RichAllWindowFunction
+import org.apache.flink.types.Row
+import org.apache.flink.util.Collector
+import org.apache.flink.streaming.api.windowing.windows.Window
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.table.functions.Accumulator
+
+import java.lang.Iterable
+import org.apache.flink.table.functions.AggregateFunction
+
+/**
+  * Computes the final aggregate value from incrementally computed aggreagtes.
+  *
+  * @param aggregates The aggregates to be computed
+  * @param aggFields the fields on which to apply the aggregate.
+  * @param forwardedFieldCount The fields to be carried from current row.
+  */
+class DataStreamProcTimeAggregateGlobalWindowFunction[W <: Window](
```

Comment:
Actually, I think the `ProcessFunction` is more extensible because the interface is more expressive and allows for more optimizations. We have to manage our own state, but that's rather a good than a bad thing because we have more control over it. That also means that we have to take care of evictions ourselves, i.e., we can have `ListState` and add values or remove them. I agree that it adds more code but not necessarily more overhead. Eviction is not free in DataStream windows, in fact it is rather expensive. Also we would need a custom `Evictor` to remove rows based on processing time.

As I said before, the `ProcessFunction` allows to continuously accumulate and retract values for aggregations. This will be more efficient than aggregating all rows for each row over and over again. Regarding integration of other SQL operators. SQL operators are assembled in a plan. A `LIMIT` or `TOP` operator will be executed after or before an aggregation but not be merged with the same operator. Moreover, a `ProcessFunction` can execute anything and more than a `WindowFunction`.